presto-main/src/main/java/com/facebook/presto/sql/analyzer/MaterializedViewQueryOptimizer.java

```diff
@@ -0,0 +1,289 @@
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupBy;
+import com.facebook.presto.sql.tree.GroupingElement;
+import com.facebook.presto.sql.tree.Identifier;
+import com.facebook.presto.sql.tree.LogicalBinaryExpression;
+import com.facebook.presto.sql.tree.Node;
+import com.facebook.presto.sql.tree.OrderBy;
+import com.facebook.presto.sql.tree.Query;
+import com.facebook.presto.sql.tree.QueryBody;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.Relation;
+import com.facebook.presto.sql.tree.Select;
+import com.facebook.presto.sql.tree.SelectItem;
+import com.facebook.presto.sql.tree.SimpleGroupBy;
+import com.facebook.presto.sql.tree.SingleColumn;
+import com.facebook.presto.sql.tree.SortItem;
+import com.facebook.presto.sql.tree.Table;
+import com.facebook.presto.sql.tree.TableSubquery;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class MaterializedViewQueryOptimizer
+        extends AstVisitor<Node, Void>
+{
+    private static final Logger logger = Logger.get(MaterializedViewQueryOptimizer.class);
+    private final Table materializedView;
+    private final ImmutableMap<String, String> baseToViewColumnMap;
+    private final Optional<Relation> baseQueryRelation;
+
+    public MaterializedViewQueryOptimizer(Table materializedView, Query originalSqlQuery)
+    {
+        this.materializedView = requireNonNull(materializedView, "materialized view is null");
+
+        QuerySpecification materializedViewDefinition = (QuerySpecification) originalSqlQuery.getQueryBody();
+        this.baseQueryRelation = materializedViewDefinition.getFrom().isPresent() ? materializedViewDefinition.getFrom() : Optional.empty();
+
+        // Create base to materialized view column map
+        ImmutableMap.Builder<String, String> baseToViewColumnMap = ImmutableMap.builder();
+
+        for (SelectItem baseTableSelectColumn : materializedViewDefinition.getSelect().getSelectItems()) {
+            if (baseTableSelectColumn instanceof SingleColumn) {
+                String baseColumnName = ((SingleColumn) baseTableSelectColumn).getExpression().toString();
+                Optional<Identifier> baseColumnAlias = ((SingleColumn) baseTableSelectColumn).getAlias();
+                String viewDerivedColumnName = baseColumnAlias.orElse(new Identifier(baseColumnName)).getValue();
+
+                baseToViewColumnMap.put(baseColumnName, viewDerivedColumnName);
+            }
+        }
+
+        this.baseToViewColumnMap = baseToViewColumnMap.build();
+    }
+
+    public Node rewrite(Node node)
+    {
+        try {
+            return process(node);
+        }
+        catch (Exception ex) {
+            logger.error(ex, "Failed to rewrite query: %s.", node);
+            return node;
+        }
```

Comment:
> When will we have this case?

This is only use to handle cases that cannot be rewritten (due to mismatch table name or column name). However, all cases will be validated before using the conversion, so this should never be triggered. I only added these lines in here to make sure all queries can be safely processed whether they can be rewritten or not.

Reply:
Just add more context, 
In case we run in to issues if the query conversion can not be done even though it has been validated, we need to ensure that we return the original query instead of aborting the query execution.